src/routes/v3/oauth_clients.rs

```diff
@@ -0,0 +1,387 @@
+    cfg.service(oauth_client_create);
+    cfg.service(oauth_client_edit);
+    cfg.service(oauth_client_delete);
+    cfg.service(get_user_clients);
+    cfg.service(get_user_oauth_authorizations);
+    cfg.service(revoke_oauth_authorization);
+}
+
+#[get("user/{user_id}/oauth_apps")]
+pub async fn get_user_clients(
+    req: HttpRequest,
+    info: web::Path<String>,
+    pool: web::Data<PgPool>,
+    redis: web::Data<RedisPool>,
+    session_queue: web::Data<AuthQueue>,
+) -> Result<HttpResponse, ApiError> {
+    let current_user = get_user_from_headers(
+        &req,
+        &**pool,
+        &redis,
+        &session_queue,
+        Some(&[Scopes::OAUTH_CLIENT_READ]),
+    )
+    .await?
+    .1;
+
+    let target_user = User::get(&info.into_inner(), &**pool, &redis).await?;
+
+    if let Some(target_user) = target_user {
+        let target_user_id: models::ids::UserId = target_user.id.into();
+        current_user.validate_can_interact_with_oauth_client(target_user_id)?;
+
+        let clients = OAuthClient::get_all_user_clients(target_user.id, &**pool).await?;
+
+        let response = clients
+            .into_iter()
+            .map(models::oauth_clients::OAuthClient::from)
+            .collect_vec();
+
+        Ok(HttpResponse::Ok().json(response))
+    } else {
+        Ok(HttpResponse::NotFound().body(""))
+    }
+}
+
+#[derive(Deserialize, Validate)]
+pub struct NewOAuthApp {
+    #[validate(
+        custom(function = "crate::util::validate::validate_name"),
+        length(min = 3, max = 255)
+    )]
+    pub name: String,
+
+    #[validate(
+        custom(function = "crate::util::validate::validate_url"),
+        length(max = 255)
+    )]
+    pub icon_url: Option<String>,
+
+    pub max_scopes: Scopes,
+
+    #[validate(length(min = 1))]
+    pub redirect_uris: Vec<String>,
+}
+
+#[post("oauth_app")]
+pub async fn oauth_client_create<'a>(
+    req: HttpRequest,
+    new_oauth_app: web::Json<NewOAuthApp>,
+    pool: web::Data<PgPool>,
+    redis: web::Data<RedisPool>,
+    session_queue: web::Data<AuthQueue>,
+) -> Result<HttpResponse, CreateError> {
+    let current_user = get_user_from_headers(
+        &req,
+        &**pool,
+        &redis,
+        &session_queue,
+        Some(&[Scopes::OAUTH_CLIENT_CREATE]),
+    )
+    .await?
+    .1;
+
+    new_oauth_app
+        .validate()
+        .map_err(|e| CreateError::ValidationError(validation_errors_to_string(e, None)))?;
+
+    let mut transaction = pool.begin().await?;
+
+    let client_id = generate_oauth_client_id(&mut transaction).await?;
+
+    let client_secret = generate_oauth_client_secret();
+    let client_secret_hash = DBOAuthClient::hash_secret(&client_secret);
+
+    let redirect_uris =
+        create_redirect_uris(&new_oauth_app.redirect_uris, client_id, &mut transaction).await?;
+
+    let client = OAuthClient {
+        id: client_id,
+        icon_url: new_oauth_app.icon_url.clone(),
+        max_scopes: new_oauth_app.max_scopes,
```

Comment:
make sure you are checking that they are not requesting restricted scopes here (and anywhere else scope issuance is involved!)

Reply:
oo good call, that would have been a big miss. I'll do a pass over all the scopes usage and add handling for that.

Reply:
When creating a client we error if restricted scopes are requested, and when issuing a token we now sutract restricted scopes just to be sure.